### PR TITLE
Nn 1832 pay endpoint

### PR DESCRIPTION
**Pull request by JamieDraperUK**
The initial version of the attendance API. It's WIP, just wanted to push a version so that John can start integrating it.

Includes the following:-

- New endpoints for attendance
- Upgraded gradle.build script to use implementation over compile
- Migrated tests to Kotlin

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -1,30 +1,47 @@
...
+apply plugin: 'kotlin'
+
 
 repositories {
     mavenCentral()
 }
 
+compileKotlin {
+    kotlinOptions {
+        jvmTarget = "1.8"
```
Comment (petergphillips) [inline build.gradle]: should be 11 I think

File: src/main/resources/application.yml [hunk 2]
```diff
@@ -25,7 +25,7 @@ spring:
       WRITE_DATES_AS_TIMESTAMPS : false
 
 server:
-  port: 8080
+  port: 8081
```
Comment (petergphillips) [inline src/main/resources/application.yml]: can this be 8082

File: src/main/resources/db/migration/V2__seed_absent_reasons.sql [hunk 3]
```diff
@@ -0,0 +1,16 @@
+DELETE FROM ABSENT_REASONS;
+
+INSERT INTO ABSENT_REASONS (ID,REASON, PNOMIS_CODE, PAID_REASON) VALUES (1, 'Acceptable absence', 'ACCAB', true);
+INSERT INTO ABSENT_REASONS (ID,REASON, PNOMIS_CODE, PAID_REASON) VALUES (2, 'Not required', 'NREQ', true);
+
+
+INSERT INTO ABSENT_REASONS (ID,REASON, PNOMIS_CODE) VALUES (3, 'Session Cancelled', 'CANC');
```
Comment (petergphillips) [inline src/main/resources/db/migration/V2__seed_absent_reasons.sql]: need to decide if all words should be capitalised or just first one

File: src/main/resources/db/migration/V1__create_offender_attendance.sql [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+  ID                              SERIAL PRIMARY KEY,
+  OFFENDER_BOOKING_ID             BIGINT          NOT NULL,
+  EVENT_ID                        BIGINT          NOT NULL,
+  EVENT_DATE                      DATE            NOT NULL,
+  EVENT_LOCATION_ID               BIGINT          NOT NULL,
+  PERIOD                          VARCHAR(2)      NOT NULL,
+  PRISON_ID                       VARCHAR(6)      NOT NULL,
+  ATTENDED                        BOOLEAN,
+  PAID                            BOOLEAN,
+  ABSENT_REASON                   VARCHAR(30),
```
Comment (Mjwillis) [inline src/main/resources/db/migration/V1__create_offender_attendance.sql]: WE'll want created by and modified on these fields, keyworker has examples of how these are set.

File: src/main/java/uk/gov/justice/digital/hmpps/whereabouts/services/AttendanceService.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+import uk.gov.justice.digital.hmpps.whereabouts.dto.AttendanceDto;
+import uk.gov.justice.digital.hmpps.whereabouts.model.Attendance;
+import uk.gov.justice.digital.hmpps.whereabouts.model.TimePeriod;
+import uk.gov.justice.digital.hmpps.whereabouts.repository.AttendanceRepository;
+
+import java.time.LocalDate;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+@Service
```
Comment (Mjwillis) [inline src/main/java/uk/gov/justice/digital/hmpps/whereabouts/services/AttendanceService.java]: Missing `@Transactional`

File: src/main/java/uk/gov/justice/digital/hmpps/whereabouts/model/Attendance.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+import lombok.NoArgsConstructor;
+import org.hibernate.validator.constraints.Length;
+
+import javax.persistence.*;
+import javax.validation.constraints.NotNull;
+import java.time.LocalDate;
+
+@Entity
+@Table(name = "OFFENDER_ATTENDANCE")
+@Data()
```
Comment (Mjwillis) [inline src/main/java/uk/gov/justice/digital/hmpps/whereabouts/model/Attendance.java]: remove `()`

File: src/test/kotlin/uk/gov/justice/digital/hmpps/whereabouts/services/OffenderEventServiceTest.kt [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+
+    private var offenderEventService: OffenderEventService? = null
+
+    @Before
+    fun setUp() {
+        offenderEventService = OffenderEventService(repository)
+    }
+
+    @Test
+    fun createOffenderEvent() {
```
Comment (andrewrlee) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/whereabouts/services/OffenderEventServiceTest.kt]: missing impl?

File: src/main/java/uk/gov/justice/digital/hmpps/whereabouts/services/Elite2ApiHealth.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+            final var responseEntity = this.restTemplate.getForEntity("/health", String.class);
+            return health(responseEntity.getStatusCode());
+        } catch (final RestClientException e) {
+            return health(HttpStatus.SERVICE_UNAVAILABLE);
+        }
+    }
+
+    private Health health(final HttpStatus code) {
+        return health (
+                Health.up(),
```
Comment (andrewrlee) [inline src/main/java/uk/gov/justice/digital/hmpps/whereabouts/services/Elite2ApiHealth.java]: Should this inspect the type of the status to determine whether it's healthy or not?

Comment (JamieDraperUK) [reply src/main/java/uk/gov/justice/digital/hmpps/whereabouts/services/Elite2ApiHealth.java]: Not sure, It's the code we use in all our backends. Will investigate.

**Pull request merged**